cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,260 @@
+ *                element i. If min_periods_col = NULL, then min_periods is used as
+ *                the static value for all elements
+ * @param[in] forward_window_col The forward window size values,
+ *                forward_window_col[i] specifies forward window size for element i.
+ *                If forward_window_col = NULL, then forward_window is used as the
+ *                static forward window size for all elements
+
+ */
+template <typename ColumnType, template <typename AggType> class agg_op, bool average>
+__global__
+void gpu_rolling(gdf_size_type nrows,
+		 ColumnType *out_col, gdf_valid_type *out_col_valid,
+		 ColumnType *in_col, gdf_valid_type *in_col_valid,
+		 gdf_size_type window,
+		 gdf_size_type min_periods,
+		 gdf_size_type forward_window,
+		 const gdf_size_type *window_col,
+		 const gdf_size_type *min_periods_col,
+		 const gdf_size_type *forward_window_col)
+{
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
+
+  gdf_size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+  gdf_size_type stride = blockDim.x * gridDim.x;
+
+  agg_op<ColumnType> op;
+
+  auto active_threads = __ballot_sync(0xffffffff, i < nrows);
+  while(i < nrows)
+  {
+    ColumnType val = agg_op<ColumnType>::IDENTITY;
+    volatile gdf_size_type count = 0;	// declare this as volatile to avoid the compiler optimizing the counter
+
+    // dynamic window handling
+    if (window_col != NULL) window = window_col[i];
+    if (min_periods_col != NULL) min_periods = max(min_periods_col[i], 1);	// at least one observation is required
+    if (forward_window_col != NULL) forward_window = forward_window_col[i];
+
+    // compute bounds
+    gdf_size_type start_index = max((gdf_size_type)0, i - window + 1);
+    gdf_size_type end_index = min(nrows, i + forward_window + 1);       // exclusive
+
+    // aggregate
+    for (size_t j = start_index; j < end_index; j++) {
+      bool const input_is_valid{gdf_is_valid(in_col_valid, j)};
+      if (input_is_valid) {
+        val = op(in_col[j], val);
+        count++;
+      }
+    }
+
+    // check if we have enough input samples
+    bool output_is_valid = (count >= min_periods);
+
+    // set the mask
+    MaskType const result_mask{__ballot_sync(active_threads, output_is_valid)};
+    gdf_index_type const out_mask_location = i / BITS_PER_MASK;
+    MaskType* const __restrict__ out_mask32 = reinterpret_cast<MaskType*>(out_col_valid);
+
+    // only one thread writes the mask
+    if (0 == threadIdx.x % warpSize)
+      out_mask32[out_mask_location] = result_mask;
+
+    // store the output value, one per thread
+    if (output_is_valid) {
+      if (average)
+        // special case for COUNT aggregator
+        out_col[i] = val / count;
+      else
+        out_col[i] = val;
+    }
+
+    // process next element 
+    i += stride;
+    active_threads = __ballot_sync(active_threads, i < nrows);
+  }
+}
+
+struct launch_kernel
```

Comment:
Either document this or give it a self-documenting name, e.g. `rolling_window_launcher`.